src/shell.js

```diff
@@ -247,7 +247,8 @@ if (ENVIRONMENT_IS_WEB || ENVIRONMENT_IS_WORKER) {
 #endif
   // blob urls look like blob:http://site.com/etc/etc and we cannot infer anything from them.
   // otherwise, slice off the final part of the url to find the script directory.
-  if (scriptDirectory.indexOf('blob:') !== 0) {
+  // be sure to leave an empty scriptDirectory alone.
+  if (scriptDirectory && scriptDirectory.indexOf('blob:') !== 0) {
```

Comment:
How about

```javascript
if (scriptDirectory.indexOf('blob:') !== 0) {
  scriptDirectory = scriptDirectory.substr(0, scriptDirectory.lastIndexOf('/')+1);
} ...
```

which reads simpler overall than the original split+slice+join+'/' sequence? (and is 130 characters vs 143 chars, and less temp garbage generation)

Reply:
Logically it seems to make sense?  If `lastIndexOf` is guaranteed to return `-1` on all browsers when the needle is not found in the haystack, then it should result in taking a zero-length sub-string `(0, 0)`.  Otherwise it should result in the path leading up to the final slash.  Adding one to the returned index then works in both cases specifically because we want to include the slash in the resulting output.  For the case where a slash is present in the string (`scriptDirectory`), this would also require that both `lastIndexOf` and `substr` operate on the same index representation, be it bytes, code points, or otherwise.

These seem like logical assumptions, but this is the exact kind of program logic that I would personally want to write property tests to cross-validate my logical argument for why the code should work, especially given non-trivial inputs and a variety of javascript implementations.

If you think that's an important thing to change here, please let me know.  But before I dive in, would you mind giving me feedback on my proposed changes to the tests?

Thanks!